lib/scala/semver/src/main/java/org/enso/semver/SemVer.java

```diff
@@ -0,0 +1,145 @@
+                    Integer.valueOf(match.group(3)),
+                    Integer.valueOf(match.group(4)),
+                    preRelease,
+                    prefix));
+
+        } else {
+            return Failure.apply(new RuntimeException("Invalid version " + version));
+        }
+    }
+
+    public static SemVer of(int major) {
+        return new SemVer(major, 0, 0);
+    }
+
+    public static SemVer of(int major, int minor) {
+        return new SemVer(major, minor, 0);
+    }
+
+    public static SemVer of(int major, int minor, int patch) {
+        return new SemVer(major, minor, patch);
+    }
+
+    public static SemVer of(int major, int minor, int patch, String preReleaseVersion) {
+        return new SemVer(major, minor, patch, Optional.of(preReleaseVersion));
+    }
+
+    public boolean isGreaterThanOrEqual(SemVer v) {
+        return major > v.major || major == v.major && (
+                minor > v.minor || minor == v.minor && patch >= v.patch);
+    }
+
+    public boolean isGreaterThan(SemVer v) {
+        return major > v.major || major == v.major && (
+                minor > v.minor || minor == v.minor && patch > v.patch);
+    }
+
+    public boolean isLessThanOrEqual(SemVer v) {
+        return major < v.major || major == v.major && (
+                minor < v.minor || minor == v.minor && patch <= v.patch);
+    }
+
+    public boolean isLessThan(SemVer v) {
+        return major < v.major || major == v.major && (
+                minor < v.minor || minor == v.minor && patch < v.patch);
+    }
+
+    public Optional<String> preReleaseVersion() {
+        return preRelease;
+    }
+
+    public int compareTo(SemVer v) {
+        if (major < v.major) return -1;
+        else if (major > v.major) return 1;
+        else {
+            if (minor < v.minor) return -1;
+            else if (minor > v.minor) return 1;
+            else {
+                if (patch < v.patch) return -1;
+                else if (patch > v.patch) return 1;
+                else {
+                    if (preRelease.isEmpty() && v.preRelease.isEmpty()) return 0;
+                    if (preRelease.isEmpty()) return 1;
+                    else if (v.preRelease.isEmpty()) return -1;
+                    else return (preRelease.equals(v.preRelease) ? 0 : preRelease.get().compareTo(v.preRelease.get()));
+                }
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        prefix.ifPresent(v -> sb.append(v));
+        sb.append(major + ".");
+        sb.append(minor + ".");
+        sb.append(patch);
+        preRelease.ifPresent(v -> sb.append("-" + v));
+        return sb.toString();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+        if (!(other instanceof SemVer)) {
+            return false;
+        }
+        return compareTo((SemVer) other) == 0;
+    }
```

Comment:
I like that we depend on `compareTo` here.

Just note that this equality ignores the `prefix`. Is that by design?

I guess it is ok to but I feel like it may be good to mention this somewhere in the docs of this class so we are all aware of this behaviour and that it was intended.

Reply:
added